ITE: drivers/clock_control: Add clock control support

bjchentc opened the pull request:
Add Clock-control(ECPM) driver module of ITE IT8xxx2 for supporting the Clock Control Interface.

Test:
Clock gating registers checking in the following platforms:
1. it8xxx2_evb -- Checked,
2. nereid -- Checked,
3. krabby/tentacruel -- Checked

Signed-off-by: BJ Chen <bj.chen@ite.com.tw>

itebj commented:
@tari 

Hi Peter,

I just found that you have a repository in the Gerrit which is WIP: nissa/nereid: enable FPU support.
I'm afraid that the FPU part would be clock gated if the option is not enabled, just shown as follows:

#ifndef CONFIG_FLOAT_HARD
        /* FPU Clock Gating */
        ecpm_regs->CGCTRL6R |= FPU_CHA_CLOCK_GATING;
#endif

If there's any further suggestion/recommendation, please kindly let me know your idea.
Thanks & regards,
BJ

tari commented:
> I'm afraid that the FPU part would be clock gated if the option is not enabled, just shown as follows:

I would add a level of indirection around gating the FPU. Add a Kconfig option:

```
config SOC_IT8XXX2_ENABLE_FPU
    bool "Enable FPU"
    default y if FPU
    help
      Enable the on-chip FPU. If disabled, the FPU will be clock gated and floating-point
      instructions will (do something, replace with what actually happens).

      If unsure, leave this enabled.
```

Then gate the clock if this symbol is disabled:

```c
#ifndef CONFIG_SOC_IT8XXX2_ENABLE_FPU
    /* FPU Clock Gating */
    ecpm_regs->CGCTRL6R |= FPU_CHA_CLOCK_GATING;
#endif
```

This way the FPU will be enabled by default if the system is built to use the FPU, but it can also be enabled if needed for some other reason (such as in my changes that use the FPU without enabling `CONFIG_FPU`).

itebj commented:
> I would add a level of indirection around gating the FPU. Add a Kconfig option:
> 
> ```
> config SOC_IT8XXX2_ENABLE_FPU
>     bool "Enable FPU"
>     default y if FPU
>     help
>       Enable the on-chip FPU. If disabled, the FPU will be clock gated and floating-point
>       instructions will (do something, replace with what actually happens).
> 
>       If unsure, leave this enabled.
> ```
> 
> Then gate the clock if this symbol is disabled:
> 
> ```c
> #ifndef CONFIG_SOC_IT8XXX2_ENABLE_FPU
>     /* FPU Clock Gating */
>     ecpm_regs->CGCTRL6R |= FPU_CHA_CLOCK_GATING;
> #endif
> ```
> 
> This way the FPU will be enabled by default if the system is built to use the FPU, but it can also be enabled if needed for some other reason (such as in my changes that use the FPU without enabling `CONFIG_FPU`).

@tari 

Truly grateful for your advice. I've modified this part as you mentioned. 
If there's any further idea, please kindly let me know your suggestions.

Thanks again.

tari commented on soc/riscv/riscv-ite/it8xxx2/Kconfig.soc:
```diff
@@ -66,6 +66,20 @@ config SOC_IT8XXX2_CPU_IDLE_GATING
+	default y if FPU
+	help
+	  Enable the on-chip FPU. If disabled, the FPU will be clock gated and floating-point
+	  instructions will (do something, replace with what actually happens).
```
Please update this to describe what actually happens if the FPU is used while disabled.

itebj replied:
Hi Peter,

Apologetic for my carelessness, I've updated it as files changed.

itebj replied:
@tari 

If there's any further suggestion, please kindly let me know.
Thanks again.

edersondisouza commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,334 @@
+#include <zephyr/dt-bindings/clock/it8xxx2_clock.h>
+#include <zephyr/drivers/clock_control/it8xxx2_clock_control.h>
+
+LOG_MODULE_REGISTER(clock_control_it8xxx2, LOG_LEVEL_ERR);
```
Why not `CONFIG_CLOCK_CONTROL_LOG_LEVEL`?

itebj replied:
> Why not `CONFIG_CLOCK_CONTROL_LOG_LEVEL`?

@edersondisouza 

Appreciative of your guidance that I forgot to replace it with the CONFIG_CLOCK_CONTROL_LOG_LEVEL, this upload should fix it. 

Thanks again.

edersondisouza commented on soc/riscv/riscv-ite/it8xxx2/Kconfig.soc:
```diff
@@ -66,6 +66,19 @@ config SOC_IT8XXX2_CPU_IDLE_GATING
+	  Enable the on-chip FPU to support the floating-point. If disabled,
+	  the FPU will be clock gated and floating-point instructions will
+	  not work because the clock would has been gated.
+
```
Why are these changes related? Maybe expand on commit message or have it a in previous patch?

itebj replied:
@edersondisouza 

The clock gating is added for the power saving when the modules are not enabled/utilized in the system, however, there's an item which is related to the FPU so I tagged Peter @tari to review this part, if the option is configured, the FPU will be available because the clock gating will not be performed in this condition.

If the CONFIG_FPU is defined, the CONFIG_SOC_IT8XXX2_ENABLE_FPU will be enabled as well, then the following actions won't be issued:
`#ifndef CONFIG_SOC_IT8XXX2_ENABLE_FPU
    /* FPU Clock Gating */
    ecpm_regs->CGCTRL6R |= FPU_CHA_CLOCK_GATING;
#endif`

tari commented on drivers/clock_control/Kconfig.it8xxx2:
```diff
@@ -0,0 +1,11 @@
+
+config CLOCK_CONTROL_IT8XXX2
+	bool "ITE IT8XXX2 clock controller driver"
+	depends on SOC_FAMILY_RISCV_ITE
```
Should this use `SOC_SERIES_RISCV32_IT8XXX2` instead, since it's a IT8xxx2 driver?

itebj replied:
@tari 

Yes, sir. You're right, I should use this `SOC_SERIES_RISCV32_IT8XXX2` instead. Now it's updated. Appreciative of your guidance.

fkokosinski commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,334 @@
+
+LOG_MODULE_REGISTER(clock_control_it8xxx2, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+#define PLLFREQ_MASK 0x0F
```
Use `BIT_MASK` here.

itebj replied:
@fkokosinski 

Got it, I'll update both of them to #define with the BIT_MASK usage.
Truly grateful.

itebj replied:
Shown as follows:
`#define PLLFREQ_MASK BIT_MASK(4)
#define MCUFREQ_MASK BIT_MASK(3)`

fkokosinski commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,334 @@
+	struct ecpm_it8xxx2_regs *const ecpm_regs = config->reg_base;
+
+	int reg_val = ecpm_regs->PLLFREQR & PLLFREQ_MASK;
+	uint8_t mcu_clock_freq_select_curr = ecpm_regs->SCDCR0 & 0x07;
```
Move `0x07` to a define (and possibly use `BIT_MASK`, if appropriate).

itebj replied:
@fkokosinski 

Thanks for guidance, and now it's updated as follows:
`uint8_t mcu_clock_freq_select_curr = ecpm_regs->SCDCR0 & MCUFREQ_MASK;`

fkokosinski commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,334 @@
+	/* Re-configure PLL clock or not. */
+	if (((ecpm_regs->PLLFREQR & 0x0F) != pll->pll_setting) ||
+	    ((ecpm_regs->SCDCR0 & 0xF0) != (pll->div_fnd << 4)) ||
+	    ((ecpm_regs->SCDCR3 & 0x0F) != pll->div_ec)) {
```
Ditto - use define and `BIT_MASK` macro or `GETMASK` macro.

itebj replied:
@fkokosinski 

Yes, sir. I've updated them as follows:

Definitions:
```
#define EC_FREQ_MASK BIT_MASK(4)
#define FNDFREQ_MASK ~(BIT_MASK(4))
```

Flow control:
```
	/* Re-configure PLL clock or not. */
	if (((ecpm_regs->PLLFREQR & PLLFREQ_MASK) != pll->pll_setting) ||
	    ((ecpm_regs->SCDCR0 & FNDFREQ_MASK) != (pll->div_fnd << 4)) ||
	    ((ecpm_regs->SCDCR3 & EC_FREQ_MASK) != pll->div_ec)) {
```

Thanks again for helps and guidance.

keith-zephyr commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,337 @@
+		ite_intc_save_and_disable_interrupts();
+	}
+	/* configure PLL/CPU/flash clock */
+	chip_configure_pll(&pll_configuration[pllfreq]);
```
You could move the PLL frequency set into `it8xxx2_clock_control_set_rate()`.  This call would then be `it8xxx2_clock_control_set_rate(dev, IT8XXX2_PLL, pllfreq);`

This would provide flexibility to the application to change the PLL and the CPU frequency.

itebj replied:
@keith-zephyr 

As mentioned in previous conversation, we'd love to add the flexibility into the code but even that the code will not open the PLL change in the real-time for design reasons. However, the flexibility should be somehow contained in the current update.

If there's any further idea/suggestion, please kindly let me know.
Thanks again for patience & guidance.

keith-zephyr commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -256,7 +263,7 @@ static void i2c_enhanced_port_set_frequency(const struct device *dev,
 	if (freq_hz) {
 		/* Get SMBus clock divide value */
-		clk_div = (IT8XXX2_ECPM_SCDCR2 & 0x0F) + 1U;
+		clk_div = (ecpm_base->SCDCR2 & 0x0F) + 1U;
```
Move this logic into the clock control driver. `clock_control_get_rate(config->clk_dev, IT8XXX2_SMB_CLOCK, &smb_clock`).

itebj replied:
Thank you, sir. I'll try my best to move this logic into clock_control_it8xxx2 module.

itebj replied:
@keith-zephyr 

This part has been updated to get the current I2C source clock settings from the API clock_control_get_rate() as follows:

```
	struct it8xxx2_clkctrl_subsys_t subsys = {
		.clk_opt = IT8XXX2_SMB,
	};
		/* Get SMBus source clock value */
	clock_control_get_rate(config->clk_dev,
		(clock_control_subsys_t *)&subsys, &i2c_src_clock);
```
After that, the further calculation of the psr is as follows:

```
	if (freq_hz) {
		/* Calculate PSR value */
		psr = (i2c_src_clock / (2U * freq_hz)) - 2U;
```

keith-zephyr commented on dts/riscv/ite/it8xxx2.dtsi:
```diff
@@ -1138,6 +1150,7 @@
 			interrupt-parent = <&intc>;
 			interrupts = <160 IRQ_TYPE_LEVEL_HIGH>;
 			status = "disabled";
+			clocks = <&ecpm IT8XXX2_ECPM_CGCTRL3R BIT(3) 1>;
```
Shouldn't this be:
```
clocks = <&ecpm IT8XXX2_ECPM_CGCTRL3R BIT(3) BIT(6)>;
```
Bit 6 of the CGCTRL3R register is marked as "always write 1 to this bit".

itebj replied:
Definitely yes, sir. My rookie mistake that I didn't modify the props field accordingly. I'll fix it ASAP.

itebj replied:
Definitely, sir. It should be BIT(6) here:
`clocks = <&ecpm IT8XXX2_ECPM_CGCTRL3R BIT(3) BIT(6)>;`

keith-zephyr commented on dts/bindings/clock/ite,it8xxx2-ecpm.yaml:
```diff
@@ -0,0 +1,49 @@
+clock-cells:
+  - offset
+  - bitmask
+  - always
```
I suggest naming "always_on_bitmask" for clarity.

itebj replied:
Got it.

galak commented on dts/bindings/clock/ite,it8xxx2-ecpm.yaml:
```diff
@@ -0,0 +1,49 @@
+    reg:
+      required: true
+
+    label:
```
remove `label` property

itebj replied:
Got it, sir. The label property has been removed.

itebj replied:
@galak

Thanks again for guidance and revisiting.
If there's any further idea, please kindly let me know.

galak commented on dts/riscv/ite/it8xxx2.dtsi:
```diff
@@ -1007,13 +1013,19 @@
-			reg = <0x00f01e00 0x11>;
+			reg = <0x00f01e00 0x16>;
 			reg-names = "ecpm";
+			label = "EC_PM";
```
remove label property

itebj replied:
Got it, sir. The label property has been removed.

itebj replied:
@galak 

Thanks again for guidance and revisiting.
If there's any further idea, please kindly let me know.

galak reviewed (changes requested):
remove label property from devicetree binding and dts files.

itebj commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,337 @@
+const struct clock_control_it8xxx2_config clock_control_it8xxx2_cfg = {
+	.reg_base = (struct ecpm_it8xxx2_regs *)DT_INST_REG_ADDR(0),
+	.pll_freq = DT_INST_PROP(0, pll_frequency),
+	.cpu_clock = DT_INST_PROP(0, mcu_clock),
```
Yes, sir. This property has been removed.

itebj commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,337 @@
+ *  the related properties and nodes will be added in the it8xxx2.dtsi and
+ *  YAML file.
+ */
+static int it8xxx2_clock_control_set_rate(const struct device *dev,
```
@keith-zephyr 

Thanks for suggestions and guidance, sir.
I've updated the code to extend the flexibility in the future but so far there're some limitations when it comes to the options of the clock_control_get_rate() and clock_control_set_rate():

1. CPU Frequency:
The Clock Frequency of the IT8XXX2 processor can be determined according to the PLL Frequency and the divisor which is defined by the MCU Clock Frequency Select field in the register SCDCR0 of ECPM. Now we still keep this part and make the sub_system more complete so that it could probably support more features in the future but by now only the clock_control_get_rate() can have the options, the clock_control_set_rate() still provides the CPU frequency configuration only, and the details are described in the next two paragraphs. 

2. PLL Frequency:
We strongly suggest that the PLL frequency NOT be configured in real-time for the PLL frequency change requires some specific flows including making the EC enter the sleep mode, that is, during the period when EC be in sleep mode, the system could be unpredictable because of a millisecond-level latency. Therefore, we only configure the PLL during the EC initialization in the very beginning, and that's why we made it to be executed in the RAM because we just want it to be faster in the initialization.

`#define __soc_ram_code __attribute__((section(".__ram_code")))`

3. I2C/SMBus Source Clock Frequency:
We also strongly suggest that the I2C/SMBus Source Clock Frequency NOT be configured in real-time.

So far the it8xxx2_clock_control_set_rate() only supports the MCU clock configuration but the source clock frequency of the I2C/SMBus. For there are further configurations in I2C/SMBus registers which would put the source clock frequency into division again accordingly, so it's worth noticing that the source clock frequency of the I2C/SMBus is different from the final Speed/Frequency of the I2C/SMBus.

The final Sspeed/Frequency can only be obtained through related I2C/SMBus registers. Besides, the change of the I2C/SMBus Source Clock Frequency could lead that the frequency won't match the definitions in the register SCLKTS_A, that is, the Speed/frequency will be wrong in such a scenario. Hence, we will still not let the configurations on I2C/SMBus Source ClockFrequency be available in options.

To sum up, the it8xxx2_clock_control_set_rate() is extended to have the options in the sub_system but unfortunately it only supports the CPU clock frequency adjustment in realtime so far. However, the modifications about the sub_system are made for the possible extensions in the future.

fkokosinski commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,442 @@
+	default:
+		LOG_ERR("Error: Specified Rate Not Supported\r\n");
+		return -ENOTSUP;
+	}
```
If the only supported rate option is `IT8XXX2_CPU`, wouldn't it make more sense to check only against this value instead of having essentially 4 branches (`IT8XXX2_PLL`, `IT8XXX2_SMB`, `IT8XXX2_SMB` and `default`) that result in the same error code being returned?

itebj replied:
Well, yes. In the original revision of code, I made it only check CPU stuffs only but now for adding flexibility as @keith-zephyr mentioned, we've added several branches for the extension in the future. Simultaneously, it also extends the options in the clock_control_get_rate().

itebj replied:
@keith-zephyr 
Got it, sir. So I've updated it to the older one which could save some code space.

fkokosinski commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,442 @@
+	ecpm_regs->CGCTRL4R |= SMB_ALL_CHANNELS_CLOCK_GATING;
+
+
+	if (!(DT_NODE_HAS_STATUS(DT_NODELABEL(usbpd0), okay))) {
```
`!okay` -> `disabled`? Would simplify this line a little bit.

itebj replied:
You're absolutely right, sir. I'll refine it ASAP.

fkokosinski commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,442 @@
+		ecpm_regs->CGCTRL5R |= PD0_CLOCK_GATING;
+	}
+
+	if (!(DT_NODE_HAS_STATUS(DT_NODELABEL(usbpd1), okay))) {
```
Ditto.

itebj replied:
Got it.

fkokosinski commented on drivers/peci/peci_ite_it8xxx2.c:
```diff
@@ -317,6 +327,13 @@ static int peci_it8xxx2_init(const struct device *dev)
+	int ret = clock_control_on(config->clk_dev,
+			       (clock_control_subsys_t *)&config->clk_cfg);
+	if (ret < 0) {
+		LOG_ERR("Turn on clock fail %d", ret);
```
`Failed to turn on clock control`?

itebj replied:
Yup, you're right, and I'll correct it as follows:

`LOG_ERR("Failed to enable clock_control_on(): %d", ret);`

fkokosinski commented on soc/riscv/riscv-ite/it8xxx2/Kconfig.soc:
```diff
@@ -66,6 +66,19 @@ config SOC_IT8XXX2_CPU_IDLE_GATING
 	  mode is always permitted.
 
+config SOC_IT8XXX2_JTAG_ENABLE
+	bool "The JTAG module clock gating is configured here"
```
`JTAG module clock gating`, no need to be more verbose here imo

itebj replied:
Go it, I'll refine it ASAP. Thanks for guidance, sir.

fkokosinski commented on soc/riscv/riscv-ite/it8xxx2/Kconfig.soc:
```diff
@@ -66,6 +66,19 @@ config SOC_IT8XXX2_CPU_IDLE_GATING
+config SOC_IT8XXX2_JTAG_ENABLE
+	bool "The JTAG module clock gating is configured here"
+	help
+	  This option determines if the JTAG module would be enabled or not.
```
`Enable support  for JTAG module gating.`

itebj replied:
Got it.

keith-zephyr reviewed (changes requested):
Just some minor style issues remain.

itebj commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,442 @@
+const uint32_t pll_reg_to_freq[8] = {MHZ(8), MHZ(16), MHZ(24), MHZ(32),
+				MHZ(48), MHZ(64), MHZ(72), MHZ(96)};
+
+const struct it8xxx2_clkctrl_freq_t mcu_clk_freq_tbl[] = {
```
Got it, sir. It has been updated as mentioned. Thanks for guidance.

itebj commented:
> Just some minor style issues remain.

@keith-zephyr 

Truly grateful for helps, suggestions and guidance.

I've updated them as the pushed upload. When it comes to the clock_control_set_rate(), I've changed it back to the older version which you mentioned that it'd save code space. 

If there's further idea/suggestion, please kindly let me know.
Thanks again.

itebj commented:
@keith-zephyr 
@fkokosinski 
@galak 

Thanks again for helps and guidance, sirs.
However, I re-pushed it for rebasing it to the upstream/main.

gmarull commented on drivers/clock_control/Kconfig.it8xxx2:
```diff
@@ -0,0 +1,11 @@
+# Copyright (c) 2022 ITE Corporation. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_ITE_IT8XXX2_CLOCK := ite,it8xxx2-ecpm
+
+config CLOCK_CONTROL_IT8XXX2
+	bool "ITE IT8XXX2 clock controller driver"
+	depends on SOC_SERIES_RISCV32_IT8XXX2
+	default $(dt_compat_enabled,$(DT_COMPAT_ITE_IT8XXX2_CLOCK))
```
please use `DT_HAS...` helpers

itebj replied:
Got it, sir, I'll fix it ASAP.

itebj replied:
Modified to be as follows:
```
config CLOCK_CONTROL_IT8XXX2
	bool "ITE IT8XXX2 Clock Control driver"
	default y
	depends on DT_HAS_ITE_IT8XXX2_ECPM_ENABLED
	help
	  Enable the ITE IT8XXX2 Clock Control Driver.
```

gmarull commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,431 @@
+	return 0;
+}
+
+const uint32_t pll_reg_to_freq[8] = {MHZ(8), MHZ(16), MHZ(24), MHZ(32),
+				MHZ(48), MHZ(64), MHZ(72), MHZ(96)};
```
please format code properly, you can try with clang-format.

itebj replied:
Thank you, sir. I'll try with the clang-format.

itebj replied:
Done with clang-format. Thanks again for the information and guidance.

gmarull commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,431 @@
+	uint8_t code = clk_cfg->bitmask;
+	uint8_t always = clk_cfg->always;
+
+	volatile uint8_t *reg =
+		(volatile uint8_t *)((uint32_t)ecpm_regs + (uint32_t)offset);
+
+	*reg |= (always | code);
```
why not use standard `sys_read8/16/32`?

itebj replied:
Truly grateful for guidance, sir. I'll try to realize it ASAP.

itebj replied:
The sys_read8 is invoked in the current revision.
Thanks for guidance, sir.

gmarull commented on drivers/clock_control/clock_control_it8xxx2.c:
```diff
@@ -0,0 +1,431 @@
+{
+	const struct clock_control_it8xxx2_config *const config = dev->config;
+	struct ecpm_it8xxx2_regs *const ecpm_regs = config->reg_base;
+	struct it8xxx2_clkctrl_subsys_t *subsys =
```
why ending `_t` a non typedef?

itebj replied:
Thanks for guidance, sir. I'll fix this ASAP.

itebj replied:
Fixed with removing the _t behind them.
Thanks again for reminding.

gmarull commented on include/zephyr/drivers/clock_control/it8xxx2_clock_control.h:
```diff
@@ -0,0 +1,78 @@
+
+#define JTAG_CHB_CLOCK_GATING BIT(3)
+#define FPU_CHA_CLOCK_GATING BIT(2)
+
+#ifndef _ASMLANGUAGE
+struct it8xxx2_clock_control_cells {
+	uint8_t offset;
+	uint8_t bitmask;
+	uint8_t always;
+};
+
+#endif
+
+
+/* Clock Control Sybsystem Options */
+enum it8xxx2_clock {
+	IT8XXX2_NULL,
+	IT8XXX2_PLL,
+	IT8XXX2_CPU,
+	IT8XXX2_SMB,
+	IT8XXX2_CLOCK_NUM_MAX = IT8XXX2_SMB,
+};
+
+struct it8xxx2_clkctrl_subsys_t {
+	uint8_t clk_opt;
+};
+
+struct it8xxx2_clkctrl_subsys_rate_t {
+	uint32_t clk_rate;
+};
+
+struct it8xxx2_clkctrl_freq_t {
+	uint8_t divisor;
+	uint32_t frequency;
+};
+
+#define ECPM_IT8XXX2_REG_BASE \
+((struct ecpm_it8xxx2_regs *)DT_REG_ADDR(DT_NODELABEL(ecpm)))
+
+void chip_pll_ctrl(enum chip_pll_mode mode);
```
is everything meant to be public API here? If so please document things properly.

itebj replied:
My bad, they shouldn't be here, these bits definitions should be somewhere else.
Thanks for correction, sir.

gmarull commented on include/zephyr/drivers/clock_control/it8xxx2_clock_control.h:
```diff
@@ -0,0 +1,78 @@
+
+#define JTAG_CHB_CLOCK_GATING BIT(3)
+#define FPU_CHA_CLOCK_GATING BIT(2)
+
+#ifndef _ASMLANGUAGE
+struct it8xxx2_clock_control_cells {
+	uint8_t offset;
+	uint8_t bitmask;
+	uint8_t always;
+};
+
+#endif
+
+
+/* Clock Control Sybsystem Options */
+enum it8xxx2_clock {
+	IT8XXX2_NULL,
+	IT8XXX2_PLL,
+	IT8XXX2_CPU,
+	IT8XXX2_SMB,
+	IT8XXX2_CLOCK_NUM_MAX = IT8XXX2_SMB,
+};
+
+struct it8xxx2_clkctrl_subsys_t {
+	uint8_t clk_opt;
+};
+
+struct it8xxx2_clkctrl_subsys_rate_t {
+	uint32_t clk_rate;
+};
+
+struct it8xxx2_clkctrl_freq_t {
+	uint8_t divisor;
+	uint32_t frequency;
+};
+
+#define ECPM_IT8XXX2_REG_BASE \
+((struct ecpm_it8xxx2_regs *)DT_REG_ADDR(DT_NODELABEL(ecpm)))
+
+void chip_pll_ctrl(enum chip_pll_mode mode);
```
looks like a driver specialization, but there's no namespace, not even a clue this is about clock control, no docs, etc.

itebj replied:
Now the definitions are moved to the chip-level registers header file as shown in the soc/riscv/riscv-ite/common/chip_chipregs.h.

Thanks again for your helps and guidance.

fkokosinski commented:
Hi @itebj, do you plan on making further changes in this PR? Because I'm not sure if this PR is ready for reviewing as it is currently, since you made some changes today. :)

itebj commented:
> Hi @itebj, do you plan on making further changes in this PR? Because I'm not sure if this PR is ready for reviewing as it is currently, since you made some changes today. :)

Hi @fkokosinski, 

Nah, I wasn't trying to change anything but there's conflicts today, so I rebased it to the upstream/main & pushed it again.

@galak 
@gmarull 
@tari 
@edersondisouza 
@keith-zephyr 

Dear sirs,

Truly grateful for suggestions all the time.
Appreciative of your guidance if you'd be available for revisiting.

Thanks again & regards,
BJ

itebj commented:
Dear sirs,

For resolving the conflicts with rebasing the latest update of the upstream/main, I've uploaded it again, please advise if there's any further suggestion/recommendation.

Thanks again & regards,
BJ

itebj commented:
> > Hi @itebj, do you plan on making further changes in this PR? Because I'm not sure if this PR is ready for reviewing as it is currently, since you made some changes today. :)
> 
> Hi @fkokosinski,
> 
> Nah, I wasn't trying to change anything but there's conflicts today, so I rebased it to the upstream/main & pushed it again.
> 
> @galak @gmarull @tari @edersondisouza @keith-zephyr
> 
> Dear sirs,
> 
> Truly grateful for suggestions all the time. Appreciative of your guidance if you'd be available for revisiting.
> 
> Thanks again & regards, BJ

Dear sirs,

If you're available, would you please revisit, we're truly appreciative of your guidance.

Thanks again & best regards, 
BJ

itebj commented:
Dear sirs,

Due to the fact that there were conflicts in days, so I rebased it again. The latest repo has been uploaded. If you're available for revisiting, we'd be truly grateful for your kindly helps.

Thanks again & best regards,
BJ

Dino-Li commented:
@galak @gmarull could you please revisit this change?

itebj commented:
Dear sirs,

For resolving conflicts, it's rebased and updated again.
If you're available for revisiting, we'd be really appreciative of your helps & guidance.

Thanks again & best regards,
BJ

itebj commented:
Dear sirs,

Happy New Year!

@galak
@gmarull
Sirs, if you'd be available, please kindly revisit, truly grateful.
Thanks again for your helps and guidance.

Warmest regards,
BJ

itebj commented:
> Dear sirs,
> 
> Happy New Year!
> 
> @galak @gmarull Sirs, if you'd be available, please kindly revisit, truly grateful. Thanks again for your helps and guidance.
> 
> Warmest regards, BJ

itebj commented:
Dear sirs,

This update is for the conflicts fby rebasing it to the latest upstream/main.
Sorry that I clicked the wrong button "Close with comment" which closed the PR, now I've reopened it.

Thanks for revisiting.

Warmest regards,
BJ

keith-zephyr commented:
It looks like there are still conflicts to resolve.

itebj commented:
> It looks like there are still conflicts to resolve.

Thank you, sir.
Now it's resolved and tested with Tentacruel.

Thanks & Best regards,
BJ

itebj commented:
@galak
@gmarull

Dear sirs, 

if you'd be available, please kindly revisit, truly appreciative of your helps.
Thanks again.

Regards,
BJ

itebj commented:
@galak
@gmarull

Dear sirs,

if you'd be available, please kindly revisit this PR.
Thanks again for guidance.

Regards,
BJ

Commit:
ITE: drivers/clock_control: Add clock control support

Add Clock-control(ECPM) driver module of ITE IT8xxx2

Test:
Clock gating registers checking in the following platforms:
1. it8xxx2_evb -- Checked,
2. nereid -- Checked,
3. tentacruel -- Checked

Signed-off-by: BJ Chen <bj.chen@ite.com.tw>

itebj commented:
> This pull request has been marked as stale because it has been open (more than) 60 days with no activity. Remove the stale label or add a comment saying that you would like to have the label removed otherwise this pull request will automatically be closed in 14 days. Note, that you can always re-open a closed pull request at any time.

Thanks for reminding and yes please help to remove the Stale label.
Truly grateful for helps.

itebj commented:
Dear sirs,

I've uploaded it again for resolving the conflicts while I'm rebasing it to the latest upstream/main.

Dear @galak and @gmarull,

Sirs, would you please re-visit for reviewing requested changes previously?
Truly appreciative of your helps.

Thanks again & Best regards,
BJ

itebj commented:
Dear sirs, 

This update fix the conflicts by rebasing it to the upstream/main. Secondly, we also move the following property from the it8xxx2.dtsi to the latest it82xx2.dtsi for the compatibility of IT82XX2:

`clocks = <&ecpm IT8XXX2_ECPM_CGCTRL4R BIT(2) 0>;`

Truly grateful for your revisiting & Best regards,
BJ

itebj commented on drivers/i2c/i2c_ite_enhance.c:
```diff
@@ -1077,6 +1089,12 @@ static const struct i2c_driver_api i2c_enhance_driver_api = {
+		.clk_cfg = {                                                    \
+			.offset = DT_INST_CLOCKS_CELL(inst, offset),            \
+			.bitmask = DT_INST_CLOCKS_CELL(inst, bitmask),          \
+			.always = DT_INST_CLOCKS_CELL(inst, always_on_bitmask)  \
```
Got it, sir. I totally agree with you, and I've added the helper macro in this upload. Thanks again for guidance.
